tool/lib/license_utils.dart

```diff
@@ -0,0 +1,278 @@
+    final String yamlString = file.readAsStringSync();
+    final YamlDocument yamlDoc = loadYamlDocument(yamlString);
+    final YamlMap yaml = yamlDoc.contents as YamlMap;
+    final YamlMap updatePaths = yaml['update_paths'];
+
+    removeLicenses = yaml['remove_licenses'];
+    addLicenses = yaml['add_licenses'];
+    includePaths = updatePaths['include'];
+    excludePaths = updatePaths['exclude'];
+    fileTypes = updatePaths['file_types'];
+  }
+
+  /// Returns the list of indices for the given [ext] of [removeLicenses]
+  /// containing the license text to remove.
+  YamlList getRemoveIndicesForExtension(String ext) {
+    final YamlMap fileType = fileTypes[_removeDotFromExtension(ext)];
+    return fileType['remove'] as YamlList;
+  }
+
+  /// Returns the index for the given [ext] of [removeLicenses] containing the
+  /// license text to remove.
+  int getAddIndexForExtension(String ext) {
+    final YamlMap fileType = fileTypes[_removeDotFromExtension(ext)];
+    return fileType['add'];
+  }
+
+  /// Returns 'true' if the file should be excluded according to the config,
+  /// 'false' otherwise.
+  bool shouldExclude(File file) {
+    bool included = false;
+    for (final includePath in includePaths) {
+      if (file.path.startsWith(includePath)) {
+        included = true;
+        break;
+      }
+    }
+    bool excluded = false;
+    for (final excludePath in excludePaths) {
+      if (file.path.startsWith(excludePath)) {
+        excluded = true;
+        break;
+      }
+    }
+    return !included || excluded;
+  }
+
+  /// The config expects an extension without a starting dot, so remove
+  /// any that might exist (e.g. in the extension that is returned by
+  /// [p.extension]).
+  String _removeDotFromExtension(String ext) {
+      if (ext.startsWith('.')) {
+        return ext.replaceFirst('.', '');
+      }
+      return ext;
+  }
+}
+
+class LicenseHeader {
+  /// If the [file] matches the given [existingLicenseText] within the first
+  /// number of [byteCount] bytes, return the 'existing_header' and
+  /// 'replacement_header' each with the stored value ('<value>'
+  /// if configured, defaults to [defaultStoreValue] or current year)
+  /// populated. For now, only up to one stored value is supported.
+  /// If the file can't be read or no match is found, throws an exception.
+  Future<Map<String, String>> getReplacementInfo(
+    File file,
+    String existingLicenseText,
+    String replacementLicenseText,
+    int byteCount, [
+    defaultStoredValue,
+  ]) async {
+    final stream = file
+        .openRead(0, byteCount)
+        .transform(utf8.decoder)
+        .handleError((e) => throw Exception(
+            'License header expected, but error reading file - $e',),);
+    await for (final content in stream) {
+      // Return just the license headers for the simple case with no stored
+      // value requested (i.e. content matches licenseText verbatim)
+      if (content.contains(existingLicenseText)) {
```

Comment:
I think this could possibly misbehave if `content` is split in the middle of a line.

I think it might be better to read the entire file (or some fixed number of bytes from the beginning of the file) using `file.readAsStringSync()` and then perform these checks.

Reply:
For large files, this could potentially cause performance slowdowns. I will keep an eye on this, but it seems to work okay for the files in our repo.